Improve error handling for `Image`, `Images`, `PreImage`, `PreImages`

ssiccha opened the pull request:
..., PreImage, and PreImages.

If one of these functions figures out that the second argument is not an
element or subset of the Range or the Source, it now tells the user.

ChrisJefferson commented:
Could you write tests to hit all these new cases, just to sanity check them?

fingolfin commented on lib/mapping.gi:
```diff
@@ -165,17 +165,22 @@ InstallGlobalFunction( Image, function ( arg )
 
-      if     FamSourceEqFamElm( FamilyObj( map ), FamilyObj( elm ) ) then
+      if not IsMapping( map ) then
+        ErrorNoReturn( "<map> must be a mapping" );
```
But why? Before this, one could have done `Image(generalMap, set)` to return the set of (generalized) images for the generalized map `generalMap`. I think that was intentionally supported.

ssiccha replied:
I think you're right. I updated the PR. The next thing to do now is to write some tests.

ssiccha commented:
I'll definitely set up some tests as @ChrisJefferson suggested and have a closer look at your inline comment @fingolfin.

ssiccha commented:
I'm updating the PR right now. I have added tests and deleted the
```
      else
          ErrorNoReturn( "<coll> must be a domain, a set, ",
                         "or a homogeneous list" );
```
line from Image. This "else" branch can't be executed since the function compares the families of the source and of the argument `elm`. The source must be a domain and thus will always have a different family in the situation that `elm` is not a homogeneous list.

ssiccha commented:
PR is updated.

ssiccha commented:
And again some HPCGAP stuff.

fingolfin commented on hpcgap/lib/mapping.gi:
```diff
@@ -195,9 +202,15 @@ InstallGlobalFunction( Image, function ( arg )
+        ErrorNoReturn( "couldn't match families of <elm> or <coll> and ",
+                       "Source(<map>),\n",
+                       "maybe <elm> or <coll> are not contained in ",
+                       "Source(<map>) or\naren't homogeneous lists," );
```
Why is that `\n` in the middle of a line, not at the end? Perhaps format it like so:
```gap
        ErrorNoReturn( "couldn't match families of <elm> or <coll> and Source(<map>),\n",
                       "maybe <elm> or <coll> are not contained in Source(<map>) or\n",
                       "aren't homogeneous lists," );
```
Or if you are concerned about line width:
```gap
        ErrorNoReturn(
            "couldn't match families of <elm> or <coll> and Source(<map>),\n",
            "maybe <elm> or <coll> are not contained in Source(<map>) or\n",
            "aren't homogeneous lists," );
```

That said, I am not even sure it's such a good idea to insert manual line breaks... ???

ssiccha replied:
Yep, you're right. I'll drop the manual line breaks. It feels a bit weird.

fingolfin commented on hpcgap/lib/mapping.gi:
```diff
@@ -195,9 +202,15 @@ InstallGlobalFunction( Image, function ( arg )
         return [];
 
+      else
+        ErrorNoReturn( "couldn't match families of <elm> or <coll> and ",
```
Why "couldn't match families" and not "families of ... don't match"?

ssiccha replied:
I'll use "families of .. don't match". Sounds better. :+1:

fingolfin commented on hpcgap/lib/mapping.gi:
```diff
@@ -247,9 +263,15 @@ InstallGlobalFunction( Images, function ( arg )
+          ErrorNoReturn( "couldn't match families of <elm> or <coll> and ",
+                         "Source(<map>),\n",
+                         "maybe <elm> or <coll> are not contained in ",
+                         "Source(<map>) or\naren't homogeneous lists," );
```
Same remarks as above.

fingolfin commented on hpcgap/lib/mapping.gi:
```diff
@@ -298,10 +326,15 @@ InstallGlobalFunction( PreImage, function ( arg )
+          ErrorNoReturn( "couldn't match families of <elm> or <coll> and ",
+                         "Range(<map>),\n",
+                         "maybe <elm> or <coll> are not contained in ",
+                         "Source(<map>) or\naren't homogeneous lists," );
```
Same as above.

ssiccha commented:
I've addressed your remarks.

wilfwilson commented on hpcgap/lib/mapping.gi:
```diff
@@ -168,17 +172,20 @@ InstallGlobalFunction( Image, function ( arg )
-           and IsSubset( Source( map ), elm ) then
+      elif CollFamSourceEqFamElms( FamilyObj( map ), FamilyObj(elm) ) then
+        if not IsSubset( Source( map ), elm ) then
+          ErrorNoReturn( "<coll> must be a subset of Source(<map>)" );
```
Why is this `<coll>`? I expect, in this context the error message to be telling me the name of a variable here. But the variable here is called `<elm>`. So I think either you should say something like `The collection <elm>...`, or in this case (that `elm` is actually a collection), you could define a new variable `coll := elm` so that you can actually refer to `coll` in a proper way.

ssiccha replied:
I'm rewriting this to `<elm>`. I'll also change the wording from `a subset of Source..` to `contained in Source..`. Talking about "subsets" feels like it implies that `elm` and `Source..` are sets.

ssiccha replied:
Done.

wilfwilson commented on hpcgap/lib/mapping.gi:
```diff
@@ -195,9 +202,15 @@ InstallGlobalFunction( Image, function ( arg )
     fi;
-    Error( "usage: Image(<map>), Image(<map>,<elm>), Image(<map>,<coll>)" );
+    ErrorNoReturn( "usage: Image(<map>), Image(<map>,<elm>), ",
+                   "Image(<map>,<coll>)" );
```
However, in my opinion, it is fine to use `<coll>` here. Here you're only giving a name for a potential argument, and in doing so suggesting that it could be a collection. You're not implying that a variable called `coll` already exists.

ssiccha replied:
Yupp. I'll keep it that way.

wilfwilson commented on hpcgap/lib/mapping.gi:
```diff
@@ -195,9 +202,15 @@ InstallGlobalFunction( Image, function ( arg )
         return [];
 
+      else
+        ErrorNoReturn( "families of <elm> or <coll> and Source(<map>) ",
```
Again, in this error message, the variable `<coll>` that it's referring to doesn't exist, so I think you should use words instead. Something like `...the families of the element or collection <elm> and Source(<map>)...`

wilfwilson commented on hpcgap/lib/mapping.gi:
```diff
@@ -222,19 +239,18 @@ InstallGlobalFunction( Images, function ( arg )
-             and IsSubset( Source( map ), elm ) then
+        elif CollFamSourceEqFamElms( FamilyObj( map ), FamilyObj(elm) ) then
+          if not IsSubset( Source( map ), elm ) then
+            ErrorNoReturn( "<coll> must be a subset of Source(<map>)" );
```
Same comments as above, I won't repeat the text but my comments will apply to all the other places where you have written `coll` in the ways that I have complained about.

wilfwilson commented on hpcgap/lib/mapping.gi:
```diff
@@ -276,16 +302,18 @@ InstallGlobalFunction( PreImage, function ( arg )
-          elif img in Range( map ) then
-            return PreImageElm( map, img );
+          if not ( IsInjective( map ) and IsSurjective( map ) ) then
+            ErrorNoReturn( "<map> must be an inj. and surj. mapping" );
```
I know this isn't your message, but I recommend removing the abbreviations, to make things as clear as possible.
```suggestion
            ErrorNoReturn( "<map> must be an injective and surjective mapping" );
```

fingolfin replied:
Actually, why not just say "bijective" (and use `IsBijective`)?

wilfwilson replied:
As well as `IsInjective` and `IsSurjective`, `IsBijective` I implies `IsTotal` and `IsSingleValued`. I don’t know, but perhaps there’s a reason we don’t want to require one or the other of those things here

fingolfin replied:
Ugh, yeah, a map from a non-trivial group to a trivial group which is only defined at `1` is an injective and surjective mapping, even single valued, but it is not total. (That's -1 points for GAP for making a very surprising change in the meaning of very well established mathematical terms, with surprising consequences *sigh*).

ssiccha replied:
Done.

ssiccha commented:
I adopted the suggested improvements.

I put the changes into a separate commit so you can see more easily what I've changed. I'll turn the new commits into a single one after the PR is accepted.

ssiccha commented:
.. and I rebased onto master.

ssiccha commented:
Squashed.
